### DISCUSSION TITLE
Approach to testing a schema definition by vetting examples?

### DISCUSSION THREAD

**Discussion by NReilingh** [cuelang/cue#866] [Q&A] [answered]
I am wondering how more experienced CUE users would approach this problem: I have created a `schema.cue` file which declares a set of keys, types, and structs that other data files can validate against, usually with `$ cue vet schema.cue datafile.cue -c -v`. This enables me to make sure my data file matches the schema I have defined.

However, I am still developing the schema, so as I make changes to it I would like to run "integration tests" of a sort that guarantee the schema stays valid for a set of model data files. I could bash script out a series of `cue vet` invocations, but I'm wondering if there is a better way to do this, perhaps using `cue test` (which I can find little documentation on), or ideally having some kind of structured output like [TAP](https://testanything.org).

**Comment by myitcv** [comment, marked as answer]
@NReilingh good question. 

[`cue test`](https://github.com/cuelang/cue/issues/209) is something that is planned but not yet implemented. 

Separately (but very much related) we have created a project called [`unity`](https://github.com/cue-sh/unity). The primary use case of `unity` is to help us build up a corpus of CUE code against which the core CUE project can run tests to ensure correctness and performance don't regress (either via `cmd/cue` or via the `cuelang.org/go/...` APIs). Projects that add themselves to the corpus write test scripts that assert various CUE behaviours, with inputs and golden file outputs (in the future we will support running Go tests in those projects that use the `cuelang.org/go/...` APIs).

Whilst the core CUE team use `unity` and the corpus to verify that there have not been any regressions in later changes, `unity` can be used in another way. CUE projects that use/rely on `cmd/cue` can use `unity` as a limited form of `cue test` (those that rely on `cuelang.org/go/...` APIs are probably better off writing Go table tests and the like). For example, consider https://github.com/cue-sh/unity-example. This project happens to be part of the `unity` corpus, but the project _also_ uses `unity` as part of its own CI to very behaviour:

https://github.com/cue-sh/unity-example/blob/941a56a0d71d43407f35fd29a3d8cfe95842a13f/.github/workflows/test.yml#L27-L28

(a step which can obviously also be run locally by a developer of the project).

First the project declares the "base" versions of CUE that it should work against with a `tests` manifest:

https://github.com/cue-sh/unity-example/blob/941a56a0d71d43407f35fd29a3d8cfe95842a13f/cue.mod/tests/tests.cue#L3

Then any number of test "scripts" can be declared:

https://github.com/cue-sh/unity-example/blob/941a56a0d71d43407f35fd29a3d8cfe95842a13f/cue.mod/tests/eval.txt

More details on the setup required, format of the test scripts etc, can be found in the [`unity` README](https://github.com/cue-sh/unity/blob/main/README.md). `unity` is relatively new, but we have an increasing number of projects adding themselves to the corpus and using it independently. 

Incidentally, having projects in the corpus is already helping us catch real bugs and smooth the release process. For the last two releases we have found 4-5 real issues this way. So adding projects to the corpus helps everyone 😄

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/866.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
